utils.py

```diff
@@ -0,0 +1,156 @@
+# more imports than we need
+import pandas as pd
+import matplotlib.pyplot as plt
+import requests
+import json
+import csv
+import io
+import numpy as np
+import pydantic
+from pydantic import BaseModel
+from datetime import datetime as dt
+from datetime import timedelta
+from typing import Tuple
+import fetch
+
+def all_submissions_df(state: str # (str: "ohio" or "michigan", i.e.)
+                          ) -> pd.DataFrame:
+    """ 
+    Takes in the desired state portal as a string and retrieves filled pd ...
+    dataframe of all portal submissions with metadata and districtr assignment
+    Note: wrapper function of fetch.py function for user-facing utils.py
+    """
+    ids_url, plans_url, cois_url, written_url, subs = submission_endpts(state)
+    plans_df, cois_df, written_df = fetch.submissions(
+                                     ids_url, plans_url, cois_url, written_url)
+    dfs = [plans_df, cois_df, written_df]
+    all_submissions = pd.concat(dfs, ignore_index=True)
+    return all_submissions
+
+def submission_dfs(state: str # (str: "ohio" or "michigan", i.e.)
+                          ) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    """ 
+    Takes in the desired state portal as a string and retrieves filled pd ...
+    dataframes for each submission type with metadata and districtr assignments
+    Note: wrapper function of fetch.py function for user-facing utils.py
+    """
+    ids_url, plans_url, cois_url, written_url, subs = submission_endpts(state)
+    plans_df, cois_df, written_df = fetch.submissions(
+                                     ids_url, plans_url, cois_url, written_url)
+    return plans_df, cois_df, written_df
+
+def all_submissions_endpts(state: str) -> Tuple[str, str]:
+    """
+    Takes in the desired state portal and returns the endpoints for all plan...
+    ids in a portal, and an endpiont for the csv for all submissions in portal
+    """
+    ids_url, plans, cois, written, all_subs_url = submission_endpts(state)
+    return ids_url, all_subs_url
+
+def submission_endpts(state: str) -> Tuple[str, str, str, str, str]:
+    """
+    Takes in the desired state portal and returns all 5 relevant api endpts
+    """
+    state = state.lower()
+    if state == "michigan":
+        ids_url = "https://o1siz7rw0c.execute-api.us-east-2.amazonaws.com/beta/submissions/districtr-ids/michigan"
```

Comment:
Ideally, strings like these wouldn't be hardcoded here (they probably belong in some sort of config/yaml file). This is totally fine for now, though.